README.md
##snapwm-stable
### it's minimal and dynamic

I started this from catwm 31/12/10 ( https://bbs.archlinux.org/viewtopic.php?id=100215&p=1 )
    See snapwm.c for thanks and licensing.
Screenshots and ramblings/updates at https://bbs.archlinux.org/viewtopic.php?id=126463


###Summary
-------


**snapwm** is a xinerama and Xrandr aware, minimal and lightweight dynamic tiling window manager.

All configuration is read from three files in ~/.config/snapwm/ . There
 are well commented samples of each file to assit in initial setting up.

*rc.conf* has colours and window manager configurations.

*key.conf* is mandatory for shortcuts and commands to run.

*apps.conf* is optional and where apps settings are read from.


###Development
------------

Enhancements and bugfixes are developed and tested in the Nextwm
 repository first.


###Modes
-----

It allows the "normal" method of tiling window managers with the new
    window as the master or with the new window opened at the bottom
    or the top of the stack.

 *There's vertical tiling mode:*

    --------------
    |        | W |
    |        |___|
    | Master |   |
    |        |___|
    |        |   |
    --------------

 *Horizontal tiling mode:*

    -------------
    |           |
    |  Master   |
    |-----------|
    | W |   |   |
    -------------

 *Grid tiling mode:*

    -------------
    |      | W  |
    |Master|    |
    |------|----|
    |      |    |
    -------------

 *Stacking mode:*

    -------------
    |   _______  |
    |  |  ___  | |
    |  | |___| | |
    |  |_______| |
    -------------


 *Fullscreen mode*(which you'll know when you see it)

 All accessible with keyboard shortcuts defined in key.conf file.
 
 * The window *W* at the top of the stack can be resized on a per desktop basis.
 * Changing a tiling mode or window size on one desktop doesn't affect the other desktops.
 * Windows can be added/removed to/from the master area with keyboard shortcuts
 * There is a bar with a desktop switcher, space to show the focused window's name and space to show external text.
 * The rc files are reloadable 'on the run'.


###Installation
------------

Need Xlib, Xinerama and Xrandr, then:

    Copy the sample.rc.conf file to $HOME/.config/snapwm/rc.conf and edit it to suit.

    Copy the sample.apps.conf file to $HOME/.config/snapwm/apps.conf and edit it to suit.

    Copy the sample.key.conf file to $HOME/.config/snapwm/key.conf and edit it to suit.

    $ make
    # make install
    $ make clean


###Output in the Bar
-----
To have the time shown in the bar using bash:

    while true; do sleep 1; line=$(date +%H:%M:%S); xsetroot -name "$line"; done


###Bugs
----

[ * No bugs for the moment ;) (I mean, no importants bugs ;)]



Changelog
27/10/17
  Fix for changing to empty desktop leaving the last focused window
   with focus
  Ignore the first Randr event
  Fix for popped windows not unmapping with client_to_desktop()

13/8/17
  Minor fixes for window mappings and code cleanup

21/8/16
  Apps started before the window manager are now managed
  Minimum window size set in rc.conf

1/7/16
  Quit if no keys file found...

9/4/16
  Added function last_win() to focus the last focused window with
   a keyboard shortcut
  Added option to sticky the focused window on unfocused desktops
   with a keyboard shortcut
  Fix for transient window bug in multiple monitor mode

23/11/15
  Fix for keyboards shortcuts using Alt failing
  Fix for window not unmapping with client_to_desktop
  Fix for popping a window back with one window in fullscreen mode

16/11/15
  Modifying key ALTSuper changed to AltSuper for consistency
  One value for DEFAULTMODE sets all desktops to that mode
  Added man page

19/9/15
  No useless gaps in fullscreen mode

29/8/15
  Fix for xsetroot fail

18/7/15
  Xrandr capabilities added. Plug'n'play monitors
  SHOW_BAR in rc.conf can now be set for each desktop
  Useless gaps can be set around windows and the bar
   in rc.conf
  Can float a window in a tiling mode or pop it back in
   with a keyboard shortcut
  Can set a window to start out floating in apps.conf
  Improvements in fullscreen mode

5/3/14
  Point update [ 1.0.6 ]
  Better window management - can cycle through all windows - transients
   included
  Minimum size for window when resizing in stacking mode
  Add new function "terminate" for keyboard shortcut to
   close any open windows, exit the window manager and shutdown(1)
   or reboot(2)
  Fix for kill_client() when window has save dialog
  Add option BAR_SHORT to rc.conf to shorten the bar's 
   length to have room for an external app
  Add option in rc.conf to select which desktop to start on
  Right click an unfocused desktop in the switcher to move
   the focused window to that desktop

6/10/13
  Point update [ 1.0.5 ]
  Option to have the bar transparent when running a compositor
  Fix for failed start when Xinerama info isn't available

17/9/13
  Point update [ 1.0.4 ]
  Better ordering of windows when changing desktops

22/4/13
  Point update
  Better Handling of mozilla products multiple destroy notify events...
  Fix for SHOW_BAR option in rc.conf not doing anything
  Marker for last focused desktop in the switcher

21/4/13
  Fix for firefox menus disappearing when moving entries

14/4/13
  Fix for window border colours not updating when the config is updated

22/3/13
  Check number of windows before anything else in prev_win
  Fix for focus not returning to current window when an app is removed from another desktop.

17/3/13
  Fix for double clicking an empty desktop in the switcher killing the xserver.
  Added check for a window manager already running.
  Option to set the the initial number of extra windows in the master area for each desktop in rc.conf.
  Add option in keys.conf for using no modifier key.
  Remove flicker in fullscreen mode when changing desktops.
  Don't change mode with windows open on update.
  Enable multimedia keys by default.

28/2/13
  Update adding fix for flicker in fullscreen mode when focusing next/previous window.
  Added option to automatically change to a tiling mode from stacking mode when 
    reaching set number of open windows.

bar.c
// bar.c [ 2.0.7 ]

static void draw_numopen(unsigned int cd, unsigned int gc);
static Drawable area_sb;
static unsigned int total_w, pos;
/* ************************** Status Bar *************************** */
void setup_status_bar() {
    unsigned int i;
    XGCValues values;

    for(i=0;i<10;++i) {
        values.foreground = theme[i].textcolor;
        values.line_width = 2;
        values.line_style = LineSolid;
        if(font.fontset)
            theme[i].gc = XCreateGC(dis, root, GCForeground|GCLineWidth|GCLineStyle,&values);
        else {
            values.font = font.font->fid;
            theme[i].gc = XCreateGC(dis, root, GCForeground|GCLineWidth|GCLineStyle|GCFont,&values);
        }
    }
    sb_desks = 0;
    for(i=0;i<DESKTOPS;++i) {
        sb_bar[i].labelwidth = wc_size(sb_bar[i].label);
        sb_bar[i].width = sb_bar[i].labelwidth + 6;
        if(sb_bar[i].width < sb_height*3/2) sb_bar[i].width = sb_height*3/2;
        sb_desks += sb_bar[i].width;
    }
    sb_desks += 2;
}

void status_bar() {
    unsigned int i, y;
    XSetWindowAttributes setattr;
    setattr.override_redirect = True;
    unsigned long vmask = CWOverrideRedirect;

    y = (topbar == 0) ? 0+ug_bar : desktops[barmon].h+bdw-ug_bar;
    sb_width = ug_bar;
    for(i=0;i<DESKTOPS;++i) {
        sb_bar[i].sb_win = XCreateSimpleWindow(dis, root, desktops[barmon].x+sb_width, y,
                                            sb_bar[i].width-2,sb_height,2,theme[3].barcolor,theme[0].barcolor);

        XSelectInput(dis, sb_bar[i].sb_win, ButtonPressMask|EnterWindowMask|LeaveWindowMask);
        XChangeWindowAttributes(dis, sb_bar[i].sb_win, vmask, &setattr);
        XMapWindow(dis, sb_bar[i].sb_win);
        sb_width += sb_bar[i].width;
    }
    sb_area = XCreateSimpleWindow(dis, root, desktops[barmon].x+sb_desks+ug_bar, y,
             desktops[barmon].w-lessbar-(sb_desks+2)-2*ug_bar,sb_height,2,theme[3].barcolor,theme[1].barcolor);

    XSelectInput(dis, sb_area, ButtonPressMask|ExposureMask|EnterWindowMask|LeaveWindowMask);
    XChangeWindowAttributes(dis, sb_area, vmask, &setattr);
    XMapWindow(dis, sb_area);
    XWindowAttributes attr;
    XGetWindowAttributes(dis, sb_area, &attr);
    total_w = attr.width;
    area_sb = XCreatePixmap(dis, root, total_w, sb_height, DefaultDepth(dis, screen));
    XFillRectangle(dis, area_sb, theme[0].gc, 0, 0, total_w, sb_height+4);
    status_text("");
    update_bar();
}

void toggle_bar() {
    if(desktops[current_desktop].screen != barmon) return;
    if(STATUS_BAR == 0) {
        if(has_bar == 0) {
            show_bar = 1;
            unmapbar();
            desktops[current_desktop].h += sb_height+4+ug_bar;
        } else {
            show_bar = 0;
            mapbar();
            desktops[current_desktop].h -= sb_height+4+ug_bar;
            update_bar();
        }

        sh = desktops[current_desktop].h;
        tile();
    }
}

void unmapbar() {
    for(int i=0;i<DESKTOPS;++i) XUnmapWindow(dis,sb_bar[i].sb_win);
    XUnmapWindow(dis, sb_area);
    has_bar = 1;
}

void mapbar() {
    for(int i=0;i<DESKTOPS;++i) XMapWindow(dis, sb_bar[i].sb_win);
    XMapWindow(dis, sb_area);
    has_bar = 0;
}

void setbaralpha() {
    unsigned int i;
    if(baralpha < 100) {
        XChangeProperty(dis, sb_area, alphaatom, XA_CARDINAL, 32, PropModeReplace, (unsigned char *) &baropacity, 1l);
        for(i=0;i<DESKTOPS;++i)
            XChangeProperty(dis, sb_bar[i].sb_win, alphaatom, XA_CARDINAL, 32, PropModeReplace, (unsigned char *) &baropacity, 1l);
    }
}

void getwindowname(Window win, unsigned int stext) {
    char *win_name;

    if(XFetchName(dis, win, &win_name) != 0) strncpy(winname, win_name, 100);
    else winname[0] = '\0';
    if(win_name) XFree(win_name);
    if(stext == 0) status_text(winname);
}

void update_bar() {
    unsigned int i;

    for(i=0;i<DESKTOPS;++i) {
        if(i != current_desktop) {
            if(desktops[i].head != NULL) {
                draw_desk(sb_bar[i].sb_win, 2, 3, (sb_bar[i].width-sb_bar[i].labelwidth)/2, sb_bar[i].label, strlen(sb_bar[i].label));
                if(showopen < 1 && desktops[i].numorder > 1) {
                    draw_numopen(i, 3);
                }
            } else {
                draw_desk(sb_bar[i].sb_win, 1, 2, (sb_bar[i].width-sb_bar[i].labelwidth)/2, sb_bar[i].label, strlen(sb_bar[i].label));
            }
            if(i == previous_desktop)
                XFillRectangle(dis, sb_bar[i].sb_win, theme[3].gc, 0, 0, 2, 2);
        } else {
            draw_desk(sb_bar[i].sb_win, 0, 1, (sb_bar[i].width-sb_bar[i].labelwidth)/2, sb_bar[i].label, strlen(sb_bar[i].label));
            if(showopen == 0 && (desktops[i].mode == 1 || desktops[i].mode == 4) && desktops[i].numorder > 1) {
                draw_numopen(i, 1);
            }
        }
    }
    if(focus != NULL) getwindowname(focus->win, 0);
    else status_text("");
}

void draw_desk(Window win, unsigned int barcolor, unsigned int gc, unsigned int x, char *string, unsigned int len) {
    XSetWindowBackground(dis, win, theme[barcolor].barcolor);
    XClearWindow(dis, win);
    if(font.fontset)
        XmbDrawString(dis, win, font.fontset, theme[gc].gc, x, font.fh, string, len);
    else
        XDrawString(dis, win, theme[gc].gc, x, font.fh, string, len);
}

void draw_numopen(unsigned int cd, unsigned int gc) {
    unsigned int i, x=0, y=sb_height-2;

    for(i=0;i<desktops[cd].numorder; ++i) {
        XFillRectangle(dis, sb_bar[cd].sb_win, theme[gc].gc, x, y, 2, 2);
        x += 3;
        if((x+3) >= sb_bar[cd].width) return;
    }
}

void draw_text(Window win, unsigned int gc, unsigned int x, char *string, unsigned int len) {
    if(font.fontset)
        XmbDrawString(dis, win, font.fontset, theme[gc].gc, x, font.fh, string, len);
    else
        XDrawString(dis, win, theme[gc].gc, x, font.fh, string, len);
}

void status_text(char *sb_text) {
    unsigned int text_length, text_start, blank_start, wsize, count = 0, wnl;
    char win_name[256];

    XFillRectangle(dis, area_sb, theme[0].gc, 0, 0, pos, sb_height+4);
    if(strlen(sb_text) < 1) sb_text = ":";
    while(sb_text[count] != '\0' && count < windownamelength) {
        win_name[count] = sb_text[count];
        ++count;
    }
    win_name[count] = '\0';
    wnl = windownamelength*font.width;
    wsize = wc_size(win_name);
    text_length = (wsize >= wnl) ? wnl : wsize;
    blank_start = wc_size(theme[mode].modename)+(4*font.width);
    pos = blank_start+wnl;
    text_start = (LA_WINDOWNAME < 1) ? blank_start : pos - text_length;

    draw_text(area_sb, 4, font.width*2, theme[mode].modename, strlen(theme[mode].modename));
    XFillRectangle(dis, area_sb, theme[wnamebg].swgc, text_start, 0, text_length, sb_height+4);
    draw_text(area_sb, 4, text_start, win_name, count);
    XCopyArea(dis, area_sb, sb_area, theme[1].gc, 0, 0, pos, sb_height+4, 0, 0);
}

void update_output(unsigned int messg) {
    unsigned int text_length=0, p_length, text_start, i, j=2, k=0;
    unsigned int wsize, n, bg=0;
    char output[256];
    char astring[256];
    char *win_name;

    if(!(XFetchName(dis, root, &win_name))) {
        strcpy(output, "&3 snapwm inc. ");
        text_length = 15;
    } else {
        while(&win_name[text_length] != '\0' && text_length < 255) {
            output[text_length] = win_name[text_length];
            ++text_length;
        }
        text_length += 1;
        output[text_length] = '\0';
    }
    if(win_name) XFree(win_name);

    for(n=0;n<text_length;++n) {
        while(output[n] == '&') {
            if(output[n+1]-'0' < 10 && output[n+1]-'0' >= 0) {
                n += 2;
            } else if(output[n+1] == 'B' && output[n+2]-'0' < 10 && output[n+2]-'0' >= 0) {
                bg = output[n+2]-'0';
                n += 3;
            } else break;
        }
        astring[k] = output[n]; ++k;
    }
    astring[k] = '\0';
    p_length = wc_size(astring);
    text_start = total_w - p_length;
    XFillRectangle(dis, area_sb, theme[0].gc, pos, 0, total_w-pos, sb_height+4);
    k = 0; // i=pos on screen k=pos in text
    for(i=text_start;i<total_w;++i) {
        if(k <= text_length) { 
            while(output[k] == '&') {
                if(output[k+1]-'0' < 10 && output[k+1]-'0' >= 0) {
                    j = output[k+1]-'0';
                    k += 2;
                } else if(output[k+1] == 'B' && output[k+2]-'0' < 10 && output[k+2]-'0' >= 0) {
                    bg = output[k+2]-'0';
                    k += 3;
                } else break;
            }
            n = 0;
            if(output[k] == '&') {
                astring[0] = '&';
                ++n;++k;
            }
            while(output[k] != '&' && output[k] != '\0' && output[k] != '\n' && output[k] != '\r') {
                astring[n] = output[k];
                ++n;++k;
            }
            if(n < 1)
                continue;
            astring[n] = '\0';
            wsize = wc_size(astring);
            XFillRectangle(dis, area_sb, theme[bg].gc, i, 0, wsize, sb_height+4);
            draw_text(area_sb, j, i, astring, n);
            i += wsize-1;
            for(n=0;n<256;++n)
                astring[n] = '\0';
        }
    }

    XCopyArea(dis, area_sb, sb_area, theme[1].gc, pos, 0, (total_w - pos), sb_height+4, pos, 0);
    for(n=0;n<256;++n)
        output[n] ='\0';
    XSync(dis, False);
    return;
}

unsigned int wc_size(char *string) {
    unsigned int num;
    XRectangle rect;
    num = strlen(string);

    if(font.fontset) {
        XmbTextExtents(font.fontset, string, num, NULL, &rect);
        return rect.width;
    } else {
        return XTextWidth(font.font, string, num);
    }
}

void bar_rt_click() {
    if(barrtclkarg.com[0] == 0) {
        unsigned int i, j=0;
        for(i=0;i<cmdcount;++i) {
            if(strncmp("barrtclickcmd", cmds[i].name, 13) == 0) {
                while(strncmp(cmds[i].list[j], "NULL", 4) != 0) {
                    barrtclkarg.com[j] = cmds[i].list[j];
                    ++j;
                }
                barrtclkarg.com[j] = NULL;
            }
        }
    }
    if(barrtclkarg.com == NULL) return;
    spawn(barrtclkarg);
    return;
}


readrc.c
// readrc.c [ 2.0.7 ]

unsigned int i, k=0, c=0;
int j=-1;

int get_value() {
    memset(dummy, 0, 256); c = 0;
    while(buffer[k] != ';' && buffer[k] != 0 && buffer[k] != '\n') {
        if(buffer[k] == '"') {
            ++k; continue;
        }
        dummy[c] = buffer[k];
        ++c; ++k;
    }
    ++k;
    if(c > 0) {
        ++c; dummy[c] = '\0';
        return 0;
    } else return 1;
}

/* *********************** Read Config File ************************ */
void read_rcfile() {
    FILE *rcfile;

    rcfile = fopen( RC_FILE, "r" );
    if (rcfile == NULL) {
        fprintf(stderr, "\033[0;34m:: readrc : \033[0;31m Couldn't find %s\033[0m \n", RC_FILE);
        return;
    } else {
        while(fgets(buffer,sizeof buffer,rcfile) != NULL) {
            /* Check for comments */
            if(buffer[0] == '#' || buffer[0] == ' ' || buffer[0] == '\n') {
                memset(buffer, '\0', 256);
                continue;
            }
            if(buffer[strlen(buffer)-1] == '\n')
                buffer[strlen(buffer)-1] = 0;
            /* Now look for info */
            if(strstr(buffer, "WINDOWTHEME" ) != NULL) {
                k = 12; // eleven letters in windowtheme plus a space
                for(i=0;i<2;++i) {
                    if(get_value() == 0) {
                        theme[i].wincolor = getcolor(dummy);
                        if(theme[i].wincolor == 1) {
                            theme[i].wincolor = getcolor(defaultwincolor[i]);
                            logger("Default Window Border Colour", "");
                        }
                    }
                }
            } else if(strstr(buffer, "RESIZEMOVEKEY" ) != NULL) {
                k = 14;
                if(get_value() == 0) {
                    if(strncmp(dummy, "Super", 5) == 0)
                        resizemovekey = Mod4Mask;
                    else
                        resizemovekey = Mod1Mask;
                }
            } else if(strstr(buffer, "DESKTOPS" ) != NULL) {
                k = 9;
                if(get_value() == 0) {
                    DESKTOPS = strtol(dummy, NULL, 10);
                    if(DESKTOPS > 12) DESKTOPS = 12;
                }
            } else if(strstr(buffer, "DEFAULT_DESK" ) != NULL) {
                k = 13;
                if(get_value() == 0)
                    default_desk = strtol(dummy, NULL, 10);
                if(default_desk > DESKTOPS) default_desk = DESKTOPS;
            } else if(strstr(buffer, "UF_WIN_ALPHA" ) != NULL) {
                k = 13;
                if(get_value() == 0) {
                    ufalpha = strtol(dummy, NULL, 10);
                    if(ufalpha < 1 || ufalpha > 100)
                        ufalpha = 100;
                }
            } else if(strstr(buffer, "BAR_ALPHA" ) != NULL) {
                k = 10;
                if(get_value() == 0) {
                    baralpha = strtol(dummy, NULL, 10);
                    if(baralpha < 1 || baralpha > 100)
                        baralpha = 100;
                }
            } else if(strstr(buffer, "CENTER_STACK" ) != NULL) {
                k = 13;
                if(get_value() == 0) {
                    if(dummy[0] == '0' || dummy[0] == '1')
                        cstack = strtol(dummy, NULL, 10);
                }
            } else if(strstr(buffer, "BORDERWIDTH" ) != NULL) {
                k = 12;
                if(get_value() == 0)
                    bdw = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "MASTERSIZE" ) != NULL) {
                k = 11;
                if(get_value() == 0)
                    msize = strtol(dummy, NULL, 10);
                if(msize > 80 || msize < 20) msize = 50;
            } else if(strstr(buffer, "ATTACHASIDE" ) != NULL) {
                k = 12;
                if(get_value() == 0) 
                    if(dummy[0] == '0' || dummy[0] == '1')
                        attachaside = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "TOPSTACK" ) != NULL) {
                k = 9;
                if(get_value() == 0)
                    if(dummy[0] == '0' || dummy[0] == '1')
                        top_stack = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "FOLLOWMOUSE" ) != NULL) {
                k = 12;
                if(get_value() == 0)
                    if(dummy[0] == '0' || dummy[0] == '1')
                        followmouse = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "LEFT_WINDOWNAME" ) != NULL) {
                k = 16;
                if(get_value() == 0)
                    if(dummy[0] == '0' || dummy[0] == '1')
                        LA_WINDOWNAME = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "CLICKTOFOCUS" ) != NULL) {
                k = 13;
                if(get_value() == 0)
                    if(dummy[0] == '0' || dummy[0] == '1')
                        clicktofocus = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "AUTO_NUM_OPEN" ) != NULL) {
                k = 14;
                if(get_value() == 0)
                    auto_num = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "AUTO_MODE" ) != NULL) {
                k = 10;
                if(get_value() == 0) {
                    k = strtol(dummy, NULL, 10);
                    if(k >= 0 && k < 5)
                        auto_mode = k;
                }
            } else if(strstr(buffer, "DEFAULTMODE" ) != NULL) {
                k = 12;
                for(i=0;i<DESKTOPS; ++i) {
                    if(get_value() == 0) {
                        j = strtol(dummy, NULL, 10);
                        if(j > -1 && j < 5)
                            desktops[i].mode = j;
                    } else desktops[i].mode = desktops[0].mode;
                }
            } else if(strstr(buffer, "NMASTER" ) != NULL) {
                k = 8;
                for(i=0;i<DESKTOPS; ++i) {
                    if(get_value() == 0) {
                        j = strtol(dummy, NULL, 10);
                        if(j > -1 && j < 10) desktops[i].nmaster = j;
                        else desktops[i].nmaster = 0;
                    }
                }
            } else if(strstr(buffer, "MINW_H" ) != NULL) {
                k = 7;
                if(get_value() == 0)
                    minww = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "MINW_W" ) != NULL) {
                k = 7;
                if(get_value() == 0)
                    minwh = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "UG_OUT" ) != NULL) {
                k = 7;
                if(get_value() == 0)
                    ug_out = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "UG_IN" ) != NULL) {
                k = 6;
                if(get_value() == 0)
                    ug_in = strtol(dummy, NULL, 10);
            } else if(strstr(buffer, "UG_BAR" ) != NULL) {
                k = 6;
                if(get_value() == 0)
                    ug_bar = strtol(dummy, NULL, 10);
            } else if(STATUS_BAR == 0) {
                if(strstr(buffer, "SWITCHERTHEME" ) != NULL) {
                    k = 14; XGCValues values;
                    for(i=0;i<5;++i) {
                        if(get_value() == 0)
                            theme[i].barcolor = (getcolor(dummy) == 1)? getcolor(defaultbarcolor[i]):getcolor(dummy);
                        values.foreground = theme[i].barcolor;
                        theme[i].swgc = XCreateGC(dis, root, GCForeground,&values);
                    }
                } else if(strstr(buffer, "STATUSTHEME" ) != NULL) {
                    k = 12;
                    for(i=0;i<10;++i) {
                        if(get_value() == 0) {
                            theme[i].textcolor = getcolor(dummy);
                            if(theme[i].textcolor == 1) {
                                theme[i].textcolor = getcolor(defaulttextcolor[i]);
                                logger("Default text colour", "");
                            }
                        }
                    }
                } else if(strstr(buffer, "BAR_MONITOR" ) != NULL) {
                    k = 12;
                    if(get_value() == 0)
                        barmonchange = strtol(dummy, NULL, 10);
                } else if(strstr(buffer, "BAR_SHORT" ) != NULL) {
                    k = 10;
                    if(get_value() == 0)
                        lessbar = strtol(dummy, NULL, 10);
                } else if(strstr(buffer, "SHOWNUMOPEN" ) != NULL) {
                    k = 12;
                    if(get_value() == 0)
                        if(dummy[0] == '0' || dummy[0] == '1')
                            showopen = strtol(dummy, NULL, 10);
                } else if(strstr(buffer, "WNAMEBG" ) != NULL) {
                    k = 8;
                    if(get_value() == 0) {
                        wnamebg = strtol(dummy, NULL, 10);
                        wnamebg = (wnamebg < 5) ? wnamebg:0;
                    }
                } else if(strstr(buffer, "TOPBAR" ) != NULL) {
                    k = 7;
                    if(get_value() == 0)
                       if(dummy[0] == '0' || dummy[0] == '1')
                           topbar = strtol(dummy, NULL, 10);
                } else if(strstr(buffer, "SHOW_BAR" ) != NULL) {
                    k = 9;
                    for(i=0;i<DESKTOPS; ++i) {
                        if(get_value() == 0) {
                            if(dummy[0] == '0')
                                desktops[i].showbar = 0;
                            else
                                desktops[i].showbar = 1;
                        } else desktops[i].showbar = 0;
                    }
                } else if(strstr(buffer, "WINDOWNAMELENGTH" ) != NULL) {
                    k = 17;
                    if(get_value() == 0)
                        windownamelength = strtol(dummy, NULL, 10);
                } else if(strstr(buffer,"FONTNAME" ) != NULL) {
                    k = 9;
                    if(get_value() == 0) {
                        memset(font_list, '\0', 256);
                        strncpy(font_list, dummy, c);
                        get_font();
                        sb_height = font.height+2;
                        font.fh = ((sb_height - font.height)/2) + font.ascent;
                    }
                } else if(strstr(buffer, "MODENAME" ) != NULL) {
                    k = 9;
                    for(i=0;i<5;++i) {
                        if(get_value() == 0)
                            theme[i].modename = strdup(dummy);
                        else
                            theme[i].modename = strdup(defaultmodename[i]);
                    }
                } else if(strstr(buffer, "DESKTOP_NAMES") !=NULL) {
                    k = 14;
                    for(i=0;i<DESKTOPS;++i) {
                        if(get_value() == 0)
                            sb_bar[i].label = strdup(dummy);
                        else {
                            if(!(defaultdesktopnames[i]))
                                sb_bar[i].label = strdup("?");
                            else sb_bar[i].label = strdup(defaultdesktopnames[i]);
                        }
                    }
                }
            }
            memset(buffer, '\0', 256);
            memset(dummy, '\0', 256);
            k=0; c=0;
        }
    }
    fclose(rcfile);
    opacity = (ufalpha/100.00) * 0xffffffff;
    baropacity = (baralpha/100.00) * 0xffffffff;
    return;
}

void get_font() {
	char *def, **missing;
	int i, n;

	missing = NULL;
	if(strlen(font_list) > 0)
	    font.fontset = XCreateFontSet(dis, (char *)font_list, &missing, &n, &def);
	if(missing) {
		if(FONTS_ERROR < 1)
            while(--n) {
                fprintf(stderr, ":: snapwm :: missing fontset: %s\n", missing[n]);
            }
		XFreeStringList(missing);
	}
	if(font.fontset) {
		XFontStruct **xfonts;
		char **font_names;

		font.ascent = font.descent = 0;
		n = XFontsOfFontSet(font.fontset, &xfonts, &font_names);
		for(i = 0, font.ascent = 0, font.descent = 0; i < n; ++i) {
			if (font.ascent < (*xfonts)->ascent) font.ascent = (*xfonts)->ascent;
            if (font.descent < (*xfonts)->descent) font.descent = (*xfonts)->descent;
			++xfonts;
		}
		font.width = XmbTextEscapement(font.fontset, " ", 1);
	} else {
		fprintf(stderr, ":: snapwm :: Font '%s' Not Found\n:: snapwm :: Trying Font 'Fixed'\n", font_list);
		if(!(font.font = XLoadQueryFont(dis, font_list))
		&& !(font.font = XLoadQueryFont(dis, "fixed")))
			fprintf(stderr, ":: snapwm :: Error, cannot load font: '%s'\n", font_list);
		font.ascent = font.font->ascent;
		font.descent = font.font->descent;
		font.width = XTextWidth(font.font, " ", 1);
	}
	font.height = font.ascent + font.descent;
}

void set_defaults() {
    unsigned int i;

    for(i=0;i<2;++i)
        theme[i].wincolor = getcolor(defaultwincolor[i]);
    if(STATUS_BAR == 0) {
        for(i=0;i<5;++i)
            theme[i].barcolor = getcolor(defaultbarcolor[i]);
        for(i=0;i<5;++i)
            theme[i].modename = strdup(defaultmodename[i]);
        for(i=0;i<12;++i) {
            if(!(defaultdesktopnames[i]))
                sb_bar[i].label = strdup("?");
            else sb_bar[i].label = strdup(defaultdesktopnames[i]);
            desktops[i].mode = mode;
        }
        for(i=0;i<10;++i)
            theme[i].textcolor = getcolor(defaulttextcolor[i]);
        strncpy(font_list, defaultfontlist, strlen(defaultfontlist));
        get_font();
        sb_height = font.height+2;
        font.fh = ((sb_height - font.height)/2) + font.ascent;
    }
    logger("\033[0;32m Setting default values", "");
    return;
}

void update_config() {
    unsigned int i, y, old_desktops = DESKTOPS, tmp = current_desktop;
    unsigned int tmpview = desktops[current_desktop].screen;

    XUngrabKey(dis, AnyKey, AnyModifier, root);
    XUngrabButton(dis, AnyButton, AnyModifier, root);
    if(font.fontset) XFreeFontSet(dis, font.fontset);
    read_rcfile();
    y = (topbar == 0) ? ug_bar : desktops[barmon].h+bdw;
    if(DESKTOPS < old_desktops) {
        save_desktop(current_desktop);
        Arg a = {.i = DESKTOPS-1};
        for(i=DESKTOPS;i<old_desktops;++i) {
            select_desktop(i);
            if(head != NULL) {
                while(desktops[current_desktop].numorder > 0) {
                    client_to_desktop(a);
                }
            }
        }
        select_desktop(tmp);
        if(current_desktop > (DESKTOPS-1)) change_desktop(a);
    }
    init_desks();
    view[tmpview].cd = tmp;
    if(STATUS_BAR == 0) {
        setup_status_bar();
        if(DESKTOPS != old_desktops) {
            for(i=0;i<old_desktops;++i)
                XDestroyWindow(dis, sb_bar[i].sb_win);
            XDestroyWindow(dis, sb_area);
            status_bar();
        } else {
            sb_width = ug_bar;
            for(i=0;i<DESKTOPS;++i) {
                XSetWindowBorder(dis,sb_bar[i].sb_win,theme[3].barcolor);
                XMoveResizeWindow(dis, sb_bar[i].sb_win, desktops[barmon].x+sb_width, y,sb_bar[i].width-2,sb_height);
                sb_width += sb_bar[i].width;
            }
            XSetWindowBorder(dis,sb_area,theme[3].barcolor);
            XSetWindowBackground(dis, sb_area, theme[1].barcolor);
            XMoveResizeWindow(dis, sb_area, desktops[barmon].x+sb_width, y, desktops[barmon].w-(sb_desks+4)+bdw-lessbar-2*ug_bar,sb_height);
            XWindowAttributes attr;
            XGetWindowAttributes(dis, sb_area, &attr);
            total_w = attr.width;
            if(area_sb != 0) {
                XFreePixmap(dis, area_sb);
            }
            area_sb = XCreatePixmap(dis, root, total_w, sb_height+4, DefaultDepth(dis, screen));
            XFillRectangle(dis, area_sb, theme[0].gc, 0, 0, total_w, sb_height+4);
        }
    }
    if(mode == 1 && head != NULL) {
        XMoveWindow(dis,current->win,current->x,2*desktops[DESKTOPS-1].h);
    }
    for(i=0;i<DESKTOPS;++i) {
        desktops[i].master_size = (desktops[i].mode == 2) ? (desktops[i].h*msize)/100 : (desktops[i].w*msize)/100;
        if(numwins < 1) {
            Arg a = {.i = desktops[current_desktop].mode};
            switch_mode(a);
        }
    }
    select_desktop(tmp);

    if(STATUS_BAR == 0) update_bar();
    for(i=0;i<num_screens;++i) {
        select_desktop(view[i].cd);
        tile();
    }
    select_desktop(tmp);
    update_current();
    setbaralpha();

    read_apps_file();
    memset(keys, 0, keycount * sizeof(key));
    memset(cmds, 0, cmdcount * sizeof(Commands));
    grabkeys();
}
